src/fldmod_by_const.jl

```diff
@@ -0,0 +1,147 @@
+@inline function unsigned_splitmul_upper(a::T, b::T) where T<:Unsigned
+    # /* split operands into halves */
+    ah,al = splitint(a)
+    bh,bl = splitint(b)
+    halfT = typeof(ah)
+    halfbits = nbits(al)
+    # /* compute partial products */
+    p0 = widemul(al, bl);
+    p1 = widemul(al, bh);
+    p2 = widemul(ah, bl);
+    p3 = widemul(ah, bh);
+    # /* sum partial products */
+    cy = ((p0 >> halfbits) + (p1%halfT) + (p2%halfT)) >> halfbits;
+    return p3 + (p2 >> halfbits) + (p1 >> halfbits) + cy;
+end
+@inline function splitmul_upper(a::T, b::T) where T<:Unsigned
+    return unsigned_splitmul_upper(a,b)
+end
+@inline function splitmul_upper(a::T, b::T) where T<:Signed
+    return signed(unsigned_splitmul_upper(unsigned(a), unsigned(b))) - ((a < 0) ? b : 0) - ((b < 0) ? a : 0);
+end
+@inline splitmul_upper(a,b) = splitmul_upper(promote(a,b)...)
+
+
+"""
+    div_by_const(x, Val(C))
+
+The idea behind `div_by_const(x, Val(C))` is that we can avoid division entirely if
+instead of dividing by `C`, we multiply by `(2^N/C)`, then "divide by" `2^N`.
+
+And if we choose `N` to be `nbits(T)`, we can avoid the "divide by `2^N`" by just taking the
+_upper half_ of the result of the multiplication. And so, to do the multiplication, we use
+a `splitwidemul` which avoids widening by doing splitting x into upper and lower bits, doing
+several smaller multiplies, and then returning the result as two Ts, for the up and lo.
+
+REQUIRES:
+  - `C` _must be_ greater than `0`
+"""
+function div_by_const(x::T, ::Val{C}) where {T, C}
+# These checks will be compiled away during specialization.
+    if C == 1
+        return x
+    elseif (C <= 0)
+        throw(DomainError("C must be > 0"))
+    end
+    inv_coeff, toshift = calculate_inv_coeff(T, C)
+    up = splitmul_upper(x, inv_coeff)
+    out = up    # By keeping only the upper half, we're essentially dividing by 2^nbits(T)
+    # This condition will be compiled away during specialization.
+    if T <: Signed
+        # Because our magic number has a leading one, the result is negative if it's Signed.
+        # We add x to give us the positive equivalent.
+        out += x
+        signshift = (nbits(x)-1)
+        signed = T(unsigned(out) >> signshift)  # "unsigned" bitshift (to read top bit)
+    end
+    out = out >> toshift
+    if T <: Signed
+        out =  out + signed     # ; leaq    (%rdx,%rax), %rax
+    end
+    return T(out)
+end
+
+"""
+    fld_by_const(x, Val(C))
+
+Like `div_by_const`, but for `fld`, not `div` -- fld rounds down for negative numbers.
+"""
+function fld_by_const(x::T, y::Val{C}) where {T<:Integer, C}
+    d = div_by_const(x,y)
+    return d - (signbit(x ⊻ C) & (d * C !=x))
+end
+
+"""
+    manual_mod(x, y, quotient)
+
+Calculate `mod(x,y)` after you've already acquired quotient, the result of `fld(x,y)`.
+"""
+@inline function manual_mod(x::T, y::T, quotient::T) where T<:Integer
+    y == -1 && return T(0)   # avoid potential overflow in fld
```

Comment:
Is -1 the only value for which fld might overflow?

Reply:
There is only one call site, and y is always != -1?  If so consider merging this into fldmod_by_const to simplify?

Reply:
Same here, i copied it from Base.mod. Yeah I should just remove the check like you're suggesting. It's already required that it be > 0.

I've added a REQUIRES note to the docstring.